### PR TITLE
Iks304 notify when metric changes to red

### PR DESCRIPTION
**Pull request by Cas-Ros**
also removed the old notification strategy

### CODE REVIEW

File: components/notifier/src/quality_time_notifier.py [hunk 1]
```diff
@@ -23,18 +23,32 @@ async def notify(log_level: int = None) -> None:
...
+                    logging.warning("unable to send the notification for %s", notification["report_uuid"])
+        else:
+            logging.info("no new red metrics")
 
         logging.info("Sleeping %.1f seconds...", sleep_duration)
         await asyncio.sleep(sleep_duration)
 
 
+def build_notification_text(text_parameters):
+    """create and format the contents of the notification"""
```
Comment (fniessink) [inline components/notifier/src/quality_time_notifier.py]:
```suggestion
    """Create and format the contents of the notification."""
```

File: components/notifier/src/quality_time_notifier.py [hunk 2]
```diff
@@ -23,18 +23,32 @@ async def notify(log_level: int = None) -> None:
             response = await session.get(f"http://{server_host}:{server_port}/api/v3/reports")
             json = await response.json()
 
-        notifications = reds_per_report(json)
+        notifications = reds_that_are_new(json)
 
+        total_new_red_metrics = 0
         for notification in notifications:
```
Comment (fniessink) [inline components/notifier/src/quality_time_notifier.py]:
```suggestion
        total_new_red_metrics = sum(notification["new_red_metrics"] for notification in notifications)
```
And line 30 can be removed.

File: components/notifier/src/quality_time_notifier.py [hunk 3]
```diff
@@ -23,18 +23,32 @@ async def notify(log_level: int = None) -> None:
...
-            text = "\n\r[" + notification["report_title"] + "](" + notification["url"] + \
-                   ") contains " + str(notification["red_metrics"]) + " red metrics"
-            send = send_notification_to_teams(notification["teams_webhook"], text)
-            if not send:
-                logging.warning("unable to send the notification for %s", notification["report_uuid"])
+            total_new_red_metrics += notification["new_red_metrics"]
+        if total_new_red_metrics > 0:
+            for notification in notifications:
+                text = build_notification_text(notifications)
```
Comment (fniessink) [inline components/notifier/src/quality_time_notifier.py]: Why are you passing all notifications to create the notification for a specific report? If report A has a web hook it means it also receives the red metrics for report B, right? That doesn't seem correct.

File: components/notifier/src/strategies/reds_that_are_new.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+                if metric["status"] == "target_not_met" and used_to_be_not_red(metric):
+                    new_red_metrics[-1]["new_red_metrics"] += 1
+
+    return new_red_metrics
+
+
+def used_to_be_not_red(metric):
+    """determine if a metric used to be something other than red"""
+    if "recent_measurements" in metric:
+        return metric["recent_measurements"][0]["count"]["status"] != "target_not_met"
```
Comment (fniessink) [inline components/notifier/src/strategies/reds_that_are_new.py]: This isn't correct: you're looking at the first recent measurement, which is the oldest recent measurement. You have to look at the last recent measurement, i.e. `metric["recent_measurements"][-1]` and check whether it is red **and** has a start and end time that are equal. You'll have to look at your unit tests, if they are passing they are probably incorrect as well...

**Review by fniessink — CHANGES_REQUESTED** [review]

File: components/notifier/src/strategies/reds_that_are_new.py (hunk 4 above)
Comment (Cas-Ros) [reply components/notifier/src/strategies/reds_that_are_new.py]: replaced the comparison between the status of the metric and the most recent measurement with a comarison between the start and end of the most recent measurement

File: components/notifier/src/quality_time_notifier.py (hunk 3 above)
Comment (Cas-Ros) [reply components/notifier/src/quality_time_notifier.py]: corrected the message to only contain 1 report, and removed the leading sentence as it's now unnecessary

**Review by fniessink — APPROVED** [review]

**Pull request merged**